### PR TITLE
Add exchange auth method + config

### PR DESCRIPTION
**Pull request by frankrowe**
## Status
**READY**

## JIRA Ticket
MOB-202

## Description
Add `SCExchangeAuthMethod`
sample config:
```
  remote: {
    auth: 'exchange',
    client_id: 'xxxxx',
    client_secret:'xxxxx',
    http_protocol: 'https',
    http_host: 'exchange.boundlessgeo.io',
    http_port: 80
  }
```

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] License

## Steps to Test or Reproduce
Outline the steps to test or reproduce the PR here.

```sh
git fetch --all
git checkout <feature_branch> 
xctool -workspace SpatialConnect.xcworkspace -scheme SpatialConnect -sdk iphonesimulator -destination 'platform=iOS Simulator,name=iPhone 6 Plus' ONLY_ACTIVE_ARCH=NO test
```

@boundlessgeo/spatial-connect

### CODE REVIEW

File: SpatialConnect/SCExchangeAuthMethod.m [hunk 1]
```diff
@@ -44,24 +43,57 @@ - (BOOL)authFromCache {
...
+  NSString *authHeader = [NSString stringWithFormat:@"Basic %@", base64Encoded];
+  NSDictionary *res = [SCHttpUtils postDataRequestAsDictBLOCKING:url
+                                                            body:authBody
+                                                            auth:authHeader
+                                                     contentType:XML];
+
+  if (res && (jwt = res[@"access_token"])) {
     [keychainItem setObject:p forKey:(__bridge id)kSecValueData];
     [keychainItem setObject:username forKey:(__bridge id)kSecAttrAccount];
+    [keychainItem setObject:jwt forKey:(__bridge id)kSecAttrGeneric];
```
Comment (georific) [inline SpatialConnect/SCExchangeAuthMethod.m]: @tetriscode @frankrowe 
Is this the best way to store the token in the keychain? or is there a better way storing the token for later use?

Comment (georific) [reply SpatialConnect/SCExchangeAuthMethod.m]: I guess I could store it in `SCCache`

File: SpatialConnect/SCAuthProtocol.h [hunk 2]
```diff
@@ -30,6 +30,12 @@
 - (BOOL)authenticate:(NSString *)u password:(NSString *)p;
 
 /*!
+ *  @brief refreshes a valid token after expiration
+ *
+ */
+- (BOOL)refreshToken;
```
Comment (mrcnc) [inline SpatialConnect/SCAuthProtocol.h]: Do you think we need to add this to the interface/protocol?  I'm thinking refreshing the token is specific to an implementation of SCAuthProtocol that uses OAuth to refresh the token.

Comment (georific) [reply SpatialConnect/SCAuthProtocol.h]: Thats cool, we can make it specific to the exchange auth

File: SpatialConnect/SCExchangeAuthMethod.m [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        NSString *e = res[@"error"];
+        DDLogInfo(@"error trying to refresh token: %@", e);
+      }
+      [self logout];
+      return false;
+    }
+    return false;
+  } @catch (NSException *e) {
+    // deal with the exception
+    DDLogInfo(@"ERROR %@", e.reason);
```
Comment (mrcnc) [inline SpatialConnect/SCExchangeAuthMethod.m]: Do you need to return false here?

Comment (georific) [reply SpatialConnect/SCExchangeAuthMethod.m]: Actually I need to remove the try catch

File: SpatialConnect/SCConfigService.m [hunk 4]
```diff
@@ -112,10 +113,15 @@ - (void)loadConfig:(SCConfig *)c {
   if (c.remote) {
     if ([c.remote.auth isEqualToString:@"no-auth"]) {
       [sc connectAuth:[[SCNoAuthMethod alloc] init]];
+    } else if ([c.remote.auth isEqualToString:@"exchange"]) {
+      [sc connectAuth:[[SCExchangeAuthMethod alloc] initWithDictionary:@{
+            @"server_url" : c.remote.httpUri,
+            @"client_id" : c.remote.clientId
```
Comment (mrcnc) [inline SpatialConnect/SCConfigService.m:119]: We shouldn't need to look for the clientId anymore so we should remove this.  My thinking is that the client id will always be `farpoint` so there is no need to define it in a config

Comment (mrcnc) [reply SpatialConnect/SCConfigService.m:119]: We can fix it in another PR...for now, keeping the client_id as a required key is fine

**Comment by mrcnc** [comment]
👍

**Pull request merged**